cranelift/frontend/src/ssa.rs

```diff
@@ -405,7 +357,69 @@ impl SSABuilder {
     /// Note: use only when you know what you are doing, this might break the SSA building problem
     pub fn remove_block_predecessor(&mut self, block: Block, inst: Inst) -> Block {
         debug_assert!(!self.is_sealed(block));
-        self.ssa_blocks[block].remove_predecessor(inst)
+        let pred = self.ssa_blocks[block].remove_predecessor(inst);
+        self.update_predecessor_cycle(block, pred, false);
+        pred
+    }
+
+    /// Call this when `pred` is not in the predecessors of `block`: before adding, or after
+    /// removing.
+    fn update_predecessor_cycle(&mut self, block: Block, mut pred: Block, add: bool) {
+        let block_ref = &self.ssa_blocks[block];
+
+        // Not counting `pred`, how many predecessors does `block` have? And will it have exactly
+        // one after the caller is done changing things?
+        let will_have_one_pred = match block_ref.predecessors.as_slice() {
+            // None: `pred` may reach `block` in a cycle. There'll be one if we're adding `pred`.
+            [] => add,
+            // One: the other predecessor may reach `block` in a cycle. There'll be one if we're
+            // removing `pred`.
+            [other] => {
+                pred = other.block;
+                !add
+            }
+            // Two or more: there is definitely no cycle through `block`, either before or after.
+            _ => return,
```

Comment:
Should this set in_predecessor_cycle to false?

Reply:
No, it must already be false: there couldn't have been a cycle before because there were at least two predecessors. That's true whether `pred` is being added or removed.

But it's an excellent question so I'm adding a `debug_assert` there to check this.

Reply:
What if you went from

```
block1:
jump block2

block2:
jump block1
```

which has a predecessor cycle to

```
block0:
jump block1

block1:
jump block2

block2:
jump block1
```

which doesn't?

Reply:
In that example, `block1` had one predecessor (`block2`), and now we're adding a second (`pred` is `block0`). This `match` is on the number of predecessors excluding `pred`, so we'd be in the one-predecessor case where `pred` gets re-assigned to point to `block2`, not this early-return case.

Reply:
I see.